### ISSUE TITLE
Error setting txhashset root to build a block: AlreadySpent

### ISSUE THREAD

**Issue by jaspervdm** [mimblewimble/grin#913]
I have been mining succesfully for a while, until suddenly the following lines kept repeating over and over:
```
Mar 29 13:10:38.696 INFO Acquiring wallet lock ...
Mar 29 13:10:38.735 INFO ... released wallet lock
Mar 29 13:10:38.736 DEBG receive_coinbase: built candidate output - Identifier(eaea643dcaab52c9f312), 50
Mar 29 13:10:38.737 DEBG receive_coinbase: BlockFees { fees: 280000000, height: 2854, key_id: Some(Identifier(eaea643dcaab52c9f312)) }
Mar 29 13:10:38.883 DEBG get_coinbase: BlockFees { fees: 280000000, height: 2854, key_id: Some(Identifier(eaea643dcaab52c9f312)) }
Mar 29 13:10:39.786 DEBG ping/pong: 167.99.87.54:13414: 17121 @ 2847 vs us: 17127 @ 2853
Mar 29 13:10:39.828 DEBG (Server ID: Port 13414) Built new block with 35 inputs and 71 outputs, network difficulty: 1, cumulative difficulty 17128
Mar 29 13:10:40.074 DEBG Error accepting peer 192.99.71.240:37316: Serialization(IOErr(Error { repr: Custom(Custom { kind: UnexpectedEof, error: StringError("failed to fill whole buffer") }) }))
Mar 29 13:10:40.319 DEBG ping/pong: 37.59.44.69:13414: 17121 @ 2847 vs us: 17127 @ 2853
Mar 29 13:10:40.708 DEBG Error returned, discarding txhashset extension.
Mar 29 13:10:40.709 ERRO Error setting txhashset root to build a block: AlreadySpent(Commitment(09579e1fbf8893eb960a899ad372dfb1d225095b666deb1ae8d48d61077f3ac22b))
Mar 29 13:10:40.709 WARN Error building new block: Chain(Other("AlreadySpent(Commitment(09579e1fbf8893eb960a899ad372dfb1d225095b666deb1ae8d48d61077f3ac22b))")). Retrying.
```

My suspicion was that it was receiving an output from the wallet that was already used in another block, hence the AlreadySpent error. Indeed, the output `09579e1f` is already used in block 2837 of testnet2. It is also spent in block 2848.

Now my two questions related to this
* The error kept coming back over and over every few seconds, clearly it kept asking the wallet for an output and getting the same one back. Would it be possible to add some logic to the wallet to automatically add the output to the wallet.dat once this error occurs? Otherwise it will get stuck in an infinte loop if it experiences this error
* I expected a wallet restore to fix this issue, since it should detect that output and add it to the `wallet.dat`, however even after a restore the problem persists. Is there maybe a bug that spent outputs are not stored in the wallet.dat file? In that case the wallet might think that specific number output isnt used yet. I can provide my wallet.seed and wallet.dat if needed

**Comment by ignopeverell** [comment]
Your first point is valid, ideally this error condition would be detected by the wallet and it would try to restore that output. Regarding your 2nd point, wallet restore doesn't work right now on fast sync'd nodes so you likely had nothing actually restored.

**Comment by jaspervdm** [comment]
I retried a `wallet restore` with an archival node, and indeed that fixed the issue, thanks!

**Comment by ignopeverell** [comment]
No worries. #927 should fix that part.

**Comment by jaspervdm** [comment]
This is fixed by #1008
